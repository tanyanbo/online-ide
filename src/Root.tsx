import React, { ReactNode } from "react";
import { Provider } from "react-redux";
import { legacy_createStore } from "redux";
import reducers from "./reducers";

let store = legacy_createStore(reducers);
export type RootState = ReturnType<typeof store.getState>;

const Root = (props: {
  children?: ReactNode;
  initialState?: RootState | {};
}) => {
  let { children, initialState } = props;
  if (initialState === undefined) initialState = {};
  store = legacy_createStore(reducers, initialState);
  return <Provider store={store}>{children}</Provider>;
};

export type AppDispatch = typeof store.dispatch;

export default Root;
