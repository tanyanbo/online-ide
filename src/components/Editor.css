.editor {
    text-align: start;
    width: 50%;
}

