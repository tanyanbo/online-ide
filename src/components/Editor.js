import React, { useEffect, useState } from "react";
import { Controlled as CodeMirror } from "react-codemirror2";
import { connect } from "react-redux";
import "codemirror/mode/javascript/javascript";
import "codemirror/mode/xml/xml";
import "codemirror/mode/css/css";
import "codemirror/mode/python/python";
import "codemirror/lib/codemirror.css";
import "codemirror/theme/material.css";
import "./Editor.css";
import { change, changeIsRunning } from "../actions";
import KEYS from "../actions/keys";

const Editor = (props) => {
  const { language, change, checked, isRunning, changeIsRunning } = props;
  const [val, setVal] = useState("");

  const onChangeCode = (editor, data, value) => {
    setVal(value);
  };

  const clearCode = () => {
    setVal("");
    props.change("", language[0]);
  };

  useEffect(() => {
    if (checked) {
      const timer = setTimeout(() => {
        change(val, KEYS[`CHANGE_${language[0]}`]);
      }, 250);
      return () => clearTimeout(timer);
    }
    if (isRunning) {
      change(val, KEYS[`CHANGE_${language[0]}`]);
      changeIsRunning(false);
    }
  }, [val, language, checked, change, isRunning, changeIsRunning]);

  return (
    <div className="editor">
      <div className="editor-top">
        <h3>{language[0]}</h3>
        <button onClick={clearCode} className="reset-button">
          &#10060;
        </button>
      </div>
      <CodeMirror
        value={val}
        options={{
          mode: language[1],
          lineNumbers: true,
          lineWrapping: true,
          lint: true,
          theme: "material",
        }}
        onBeforeChange={onChangeCode}
      />
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    checked: state.checkbox.checked,
    isRunning: state.isRunning.isRunning,
  };
};

export default connect(mapStateToProps, {
  change,
  changeIsRunning,
})(Editor);
