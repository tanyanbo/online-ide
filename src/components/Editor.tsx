import React, { useEffect, useState } from "react";
import { Controlled as CodeMirror } from "react-codemirror2";
import { connect, ConnectedProps } from "react-redux";
import "codemirror/mode/javascript/javascript";
import "codemirror/mode/xml/xml";
import "codemirror/mode/css/css";
import "codemirror/mode/python/python";
import "codemirror/lib/codemirror.css";
import "codemirror/theme/material.css";
import "./Editor.css";
import { change, changeIsRunning } from "../actions";
import KEYS from "../actions/keys";
import { RootState } from "../index";

const mapStateToProps = (state: RootState) => {
  return {
    checked: state.checkbox.checked,
    isRunning: state.isRunning.isRunning,
  };
};

const connector = connect(mapStateToProps, { change, changeIsRunning });

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = PropsFromRedux & {
  language: readonly string[];
  checked: boolean;
  isRunning: boolean;
  hasBar?: boolean;
  handleMouseDown?: () => void;
  pos: string;
};

const Editor = React.forwardRef((props: Props, ref) => {
  const {
    language,
    change,
    checked,
    isRunning,
    changeIsRunning,
    hasBar,
    handleMouseDown,
    pos,
  } = props;
  const [code, setCode] = useState<string>("");

  const onChangeCode = (editor: any, data: any, value: string) => {
    setCode(value);
  };

  const clearCode = () => {
    setCode("");
    change("", KEYS[`CHANGE_${pos}`]);
  };

  useEffect(() => {
    if (checked) {
      const timer = setTimeout(() => {
        change(code, KEYS[`CHANGE_${pos}`]);
      }, 250);
      return () => clearTimeout(timer);
    }
    if (isRunning) {
      change(code, KEYS[`CHANGE_${pos}`]);
      changeIsRunning(false);
    }
  }, [code, language, pos, checked, change, isRunning, changeIsRunning]);

  return (
    <div className={hasBar ? "editor-with-bar" : "editor"} ref={ref}>
      {hasBar && (
        <div className="resize-bar" onMouseDown={handleMouseDown}></div>
      )}
      <div className="editor-top">
        <h3>{language[0]}</h3>
        <button onClick={clearCode} className="reset-button">
          &#10060;
        </button>
      </div>
      <CodeMirror
        className="code-mirror"
        value={code}
        options={{
          mode: language[1],
          lineNumbers: true,
          lineWrapping: true,
          lint: true,
          theme: "material",
          dragDrop: false,
          matchBrackets: true,
        }}
        onBeforeChange={onChangeCode}
      />
    </div>
  );
});

export default connector(Editor);
