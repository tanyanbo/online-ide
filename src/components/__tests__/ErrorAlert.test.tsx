import { fireEvent, render, screen } from "@testing-library/react";
import Root from "../../Root";
import Editor from "../Editor";
import LANGUAGES from "../../shared/Languages";
import React from "react";
import renderer, { act } from "react-test-renderer";

describe("<ErrorAlert />", () => {
  it("shows red exclamation mark when there is an error in the javascript code", () => {
    render(
      <Root initialState={{ srcDoc: { js: "console.log(a)" } }}>
        <Editor pos={"3"} hasBar={true} language={LANGUAGES.JS} />
      </Root>
    );

    expect(screen.getByText("!")).toBeInTheDocument();
  });

  it("does not show red exclamation mark when there are no errors in the javascript code", () => {
    render(
      <Root initialState={{ srcDoc: { js: "console.log(1)" } }}>
        <Editor pos={"3"} hasBar={true} language={LANGUAGES.JS} />
      </Root>
    );

    expect(screen.queryByText("!")).not.toBeInTheDocument();
  });

  it("alerts the right message after clicking on exclamation mark: javascript", () => {
    render(
      <Root initialState={{ srcDoc: { js: "console.log(a)" } }}>
        <Editor pos={"3"} hasBar={true} language={LANGUAGES.JS} />
      </Root>
    );

    const alertMock = jest.spyOn(window, "alert").mockImplementation();
    fireEvent.click(screen.getByText("!"));
    expect(alertMock).toBeCalledWith("a is not defined");
  });

  it("shows red exclamation mark when there is an error in the typescript code", async () => {
    let component;
    await act(async () => {
      component = renderer.create(
        <Root
          initialState={{
            srcDoc: {
              js: `const b: number = 10
            console.log(a)`,
            },
            languages: { languages: "HTML+CSS+TS" },
          }}
        >
          <Editor pos={"3"} hasBar={true} language={LANGUAGES.TS} />
        </Root>
      );
    });

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("does not show red exclamation mark when there are no errors in the typescript code", async () => {
    let component;
    await act(async () => {
      component = renderer.create(
        <Root
          initialState={{
            srcDoc: {
              js: `const b: number = 10
            console.log(b)`,
            },
            languages: { languages: "HTML+CSS+TS" },
          }}
        >
          <Editor pos={"3"} hasBar={true} language={LANGUAGES.TS} />
        </Root>
      );
    });

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("alerts the right message after clicking on exclamation mark: typescript", async () => {
    let component;
    await act(async () => {
      component = renderer.create(
        <Root
          initialState={{
            srcDoc: {
              js: `const b: number = 10
            console.log(a)`,
            },
            languages: { languages: "HTML+CSS+TS" },
          }}
        >
          <Editor pos={"3"} hasBar={true} language={LANGUAGES.TS} />
        </Root>
      );
    });

    const alertMock = jest.spyOn(window, "alert").mockImplementation();
    const button = await component.root.findByProps({
      className: "error-wrap",
    });
    button.props.onClick();

    expect(alertMock).toBeCalledWith("a is not defined");
  });
});
