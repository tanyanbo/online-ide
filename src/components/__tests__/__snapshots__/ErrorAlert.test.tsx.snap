// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorAlert /> alerts the right message where there is an error in the typescript code 1`] = `
<div
  className="editor-with-bar"
>
  <div
    className="resize-bar"
  />
  <div
    className="editor-top"
  >
    <h3>
      TS
    </h3>
    <button
      className="reset-button"
      onClick={[Function]}
    >
      ❌
    </button>
  </div>
  <div
    className="react-codemirror2 code-mirror"
  />
  <div
    className="error-wrap"
    onClick={[Function]}
  >
    !
  </div>
</div>
`;

exports[`<ErrorAlert /> does not show red exclamation mark when there are no errors in the typescript code 1`] = `
<div
  className="editor-with-bar"
>
  <div
    className="resize-bar"
  />
  <div
    className="editor-top"
  >
    <h3>
      TS
    </h3>
    <button
      className="reset-button"
      onClick={[Function]}
    >
      ❌
    </button>
  </div>
  <div
    className="react-codemirror2 code-mirror"
  />
</div>
`;

exports[`<ErrorAlert /> shows red exclamation mark when there is an error in the typescript code 1`] = `
<div
  className="editor-with-bar"
>
  <div
    className="resize-bar"
  />
  <div
    className="editor-top"
  >
    <h3>
      TS
    </h3>
    <button
      className="reset-button"
      onClick={[Function]}
    >
      ❌
    </button>
  </div>
  <div
    className="react-codemirror2 code-mirror"
  />
  <div
    className="error-wrap"
    onClick={[Function]}
  >
    !
  </div>
</div>
`;
